src/scene/forward-renderer.js

```diff
@@ -823,6 +828,143 @@ class ForwardRenderer {
         }
     }
 
+    // placeholder LUT textures for area light
+    _createAreaLightPlaceholderLuts() {
+        var placeholderLutTex =  new Texture(this.device, {
+            width: 2,
+            height: 2,
+            format: PIXELFORMAT_R8_G8_B8_A8
+        });
+        placeholderLutTex.name = 'placeholder';
+
+        var pixels = placeholderLutTex.lock();
+        for (var i = 0; i < 4; i++) {
+            for (var c = 0; c < 4; c++) {
+                pixels[i * 4 + c] = 0;
+            }
+        }
+        placeholderLutTex.unlock();
+
+        this.device.scope.resolve('areaLightsLutTex1').setValue(placeholderLutTex);
+        this.device.scope.resolve('areaLightsLutTex2').setValue(placeholderLutTex);
+    }
+
+    // creates LUT texture used by area lights
+    _uploadAreaLightLuts(resource) {
+
+        var versions = new Int16Array(resource, 32768 * 4, 2);
+
+        this._areaLightLuts = {
+            data1: new Float32Array(resource, 0, 16384),
+            data2: new Float32Array(resource, 16384 * 4, 16384),
+            majorVersion: versions[0],
+            minorVersion: versions[1]
+        };
+
+        function createTexture(device, data, format) {
+            var tex = new Texture(device, {
+                width: 64,
+                height: 64,
+                format: format,
+                addressU: ADDRESS_CLAMP_TO_EDGE,
+                addressV: ADDRESS_CLAMP_TO_EDGE,
+                type: TEXTURETYPE_DEFAULT,
+                magFilter: FILTER_LINEAR,
+                minFilter: FILTER_NEAREST,
+                anisotropy: 1
+            });
+
+            tex.lock().set(data);
+            tex.unlock();
+            tex.upload();
+
+            return tex;
+        }
+
+        function offsetScale(data, offset, scale) {
+
+            var count = data.length;
+            var ret = new Float32Array(count);
+            for (var i = 0; i < count; i++) {
+                var n = i % 4;
+                ret[i] = (data[i] + offset[n]) * scale[n];
+            }
+            return ret;
+        }
+
+        function convertToHalfFloat(data) {
+
+            var count = data.length;
+            var ret = new Uint16Array(count);
+            var float2Half = math.float2Half;
+            for (var i = 0; i < count; i++) {
+                ret[i] = float2Half(data[i]);
+            }
+
+            return ret;
+        }
+
+        function convertToUint(data){
+
+            var count = data.length;
+            var ret = new Uint8ClampedArray(count);
+            for (var i = 0; i < count; i++) {
+                ret[i] = data[i] * 255;
+            }
+
+            return ret;
+        }
+
+        // create texture if not already
+        var luts = this._areaLightLuts;
+        if (luts && !luts.ready) {
+            if (luts.majorVersion !== 0 || luts.minorVersion !== 1) {
+                console.warn(`areaLightData version: ${luts.version} is not yet supported!`);
```

Comment:
in this case we should stop processing data below, as it might crash ..

Reply:
makes sense - I'll make this more robust

Reply:
also I'm not sure ${luts.version} exists?

Reply:
yeah that's a typo :(

Reply:
fixed